// @operational/components/lib/DataTable/DataTable.styled.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const styled_1 = __importDefault(require("../utils/styled"));
const DataTable_util_1 = require("./DataTable.util");
exports.Container = styled_1.default("div", { shouldForwardProp: prop => prop !== "width" }) `
  width: ${({ width }) => width};
  overflow: visible;
`;
exports.HeadersContainer = styled_1.default("div") `
  display: grid;
  grid-template-columns: repeat(${({ numColumns, columnWidth }) => `${numColumns}, ${columnWidth}`});
  grid-template-rows: repeat(${({ numHeaders, rowHeight }) => `${numHeaders}, ${DataTable_util_1.getHeaderRowHeight(rowHeight)}px`});
`;
exports.Row = styled_1.default("div") `
  display: grid;
  grid-template-columns: repeat(${({ numCells, cellWidth }) => `${numCells}, ${cellWidth}`});
`;
exports.Cell = styled_1.default("div", { shouldForwardProp: prop => !["isEvenRow", "height", "cell"].includes(prop) }) `
  position: relative;
  display: flex;
  align-items: center;
  border-top: 0;
  border-left: ${({ cell }) => (cell === 1 ? "1px solid" : 0)};
  border-bottom: 1px solid;
  border-right: 1px solid;
  border-color: ${({ theme }) => theme.color.border.medium};
  height: ${({ height }) => height}px;
  font-family: ${({ theme }) => theme.font.family.code};
  font-weight: ${({ theme }) => theme.font.weight.regular};
  padding: 0 ${({ theme }) => theme.space.content}px;
  color: ${({ theme }) => theme.color.text.default};
  grid-column: ${({ cell }) => cell};
  background-color: ${({ theme, isEvenRow }) => (isEvenRow ? theme.color.background.almostWhite : theme.color.white)};
`;
exports.HeaderRow = styled_1.default("div") `
  left: 0;
  width: 100%;
  position: sticky;
  grid-row: 1;
  top: 0;
  z-index: 100;
  height: ${({ rowHeight }) => DataTable_util_1.getHeaderRowHeight(rowHeight)}px;
`;
exports.HeaderCell = styled_1.default(exports.Cell) `
  position: relative;
  background-color: ${({ theme }) => theme.color.background.light};
  color: ${({ theme }) => theme.color.text.dark};
  font-weight: ${({ theme }) => theme.font.weight.bold};
  border-top: ${({ rowIndex }) => (rowIndex === 0 ? "1px solid" : 0)};
  border-color: ${({ theme }) => theme.color.border.default};
  height: ${({ rowHeight }) => DataTable_util_1.getHeaderRowHeight(rowHeight)}px;
`;
exports.DataWrapper = styled_1.default("div") `
  position: absolute;
  width: 100%;
  top: ${({ numHeaders, rowHeight }) => numHeaders * DataTable_util_1.getHeaderRowHeight(rowHeight)}px;
`;
