// @operational/components/lib/Progress/Progress.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@emotion/core");
const React = __importStar(require("react"));
const styled_1 = __importDefault(require("../utils/styled"));
const Container = styled_1.default("div")({
    label: "progress",
    width: "100%",
    overflowX: "hidden",
    textAlign: "center",
    left: 0,
    position: "fixed",
    backgroundColor: "transparent",
}, ({ theme, bottom }) => ({
    zIndex: theme.zIndex.globalProgress,
    top: bottom ? "auto" : 0,
    bottom: bottom ? 0 : "auto",
}));
const fillProgress = core_1.keyframes({
    from: {
        transform: "translate3d(-100%, 0, 0)",
    },
    to: {
        transform: "translate3d(0, 0, 0)",
    },
});
const Bar = styled_1.default("div")(({ theme }) => ({
    width: "100%",
    height: 3,
    backgroundColor: theme.color.primary,
    animation: `${fillProgress} cubic-bezier(0, 0.9, 0.26, 1) forwards 20s`,
}));
const Progress = (_a) => {
    var { onRetry, onClose } = _a, props = __rest(_a, ["onRetry", "onClose"]);
    return (React.createElement(Container, Object.assign({}, props),
        React.createElement(Bar, null)));
};
exports.default = Progress;
